pkg/provider/aws/iam.go

```diff
@@ -0,0 +1,86 @@
+package aws
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/iam"
+	awsv1alpha1 "github.com/openshift/aws-account-operator/pkg/apis/aws/v1alpha1"
+)
+
+type awsStatement struct {
+	Effect    string                 `json:"Effect"`
+	Action    []string               `json:"Action"`
+	Resource  []string               `json:"Resource,omitempty"`
+	Condition *awsv1alpha1.Condition `json:"Condition,omitempty"`
+	Principal *awsv1alpha1.Principal `json:"Principal,omitempty"`
+}
+
+type policyDoc struct {
+	Version   string         `json:"Version"`
+	Statement []awsStatement `json:"Statement"`
+}
+
+func RefreshIAMPolicy(awsClient Client, federatedRole *awsv1alpha1.AWSFederatedRole, awsAccountID, uid string) error {
+	roleName := federatedRole.Name + "-" + uid
+	policyName := federatedRole.Spec.AWSCustomPolicy.Name + "-" + uid
+	policyArn := fmt.Sprintf("arn:aws:iam::%s:policy/%s", awsAccountID, policyName)
+
+	statements := make([]awsStatement, 0, len(federatedRole.Spec.AWSCustomPolicy.Statements))
+	for _, sm := range federatedRole.Spec.AWSCustomPolicy.Statements {
+		statement := awsStatement(sm)
+		statements = append(statements, statement)
+	}
+
+	pd := policyDoc{
+		Version:   "2012-10-17",
+		Statement: statements,
+	}
+
+	payload, err := json.Marshal(&pd)
+	if err != nil {
+		return err
+	}
+
+	if _, err := awsClient.DetachRolePolicy(&iam.DetachRolePolicyInput{
+		PolicyArn: &policyArn,
+		RoleName:  &roleName,
+	}); err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() != iam.ErrCodeNoSuchEntityException {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+
+	if _, err := awsClient.DeletePolicy(&iam.DeletePolicyInput{
+		PolicyArn: &policyArn,
+	}); err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() != iam.ErrCodeNoSuchEntityException {
+				return err
+			}
+		} else {
+			return err
+		}
+	}
+
+	if _, err := awsClient.CreatePolicy(&iam.CreatePolicyInput{
```

Comment:
we may need some sort of retry with exponential back off in the case that the API gives the EntityAlreadyExists err.

Reply:
Do you mean backoff retry? But I see we are doing a different thing here https://github.com/openshift/aws-account-operator/blob/master/pkg/controller/awsfederatedaccountaccess/awsfederatedaccountaccess_controller.go#L431